### PR TITLE
Moved socket initialization to wroker processes if SO_REUSEPORT is available

### PR DESCRIPTION
**Pull request by dmatetelki**
Many tests are failing.

### CODE REVIEW

**Comment by dmatetelki** [comment]
Explained at #142

**Comment by dmatetelki** [comment]
The tests fail because `hitch_test.sh` looks for `hitch_hosts()` with `hitch_pid()`. The parent process is not the process we look for now, but the children processes.

**Comment by dmatetelki** [comment]
I've assumed that on linux `SO_REUSEPORT_WORKS` is automatically enabled, but on the tavis machine that does not seem like to be the case.

File: src/tests/hitch_test.sh [hunk 1]
```diff
@@ -182,15 +182,25 @@ hitch_pid() {
...
 
 #-
-# Usage: hitch_hosts
+# Usage: hitch_worker_pids
 #
-# Print a list of hosts for the daemon started with `start_hitch`, usually in
+# Print the PIDs of the children of the daemon started with `start_hitch`
+
+hitch_worker_pids() {
+	ps --ppid "$(hitch_pid)" | grep -Eo  "^[ ]*[0-9]+"
```
Comment (klenan) [inline src/tests/hitch_test.sh]: OpenBSD's [ps(1)](https://man.openbsd.org/ps) doesn't know about `--ppid`, but [pgrep(1)](https://man.openbsd.org/pgrep) does via the short option `-P`, which behaves the same as FreeBSD's pgrep(1) and Debian's [pgrep(1)](https://manpages.debian.org/pgrep).

As simple

     pgrep -P "$(hitch_pid)"

is therefore portable and much simpler as it already lists child PIDs one per line.

File: src/tests/hitch_test.sh [hunk 2]
```diff
@@ -221,6 +231,42 @@ hitch_hosts() {
...
+#-
+# Usage: hitch_hosts
+#
+# Print a list of hosts of the parent or children processes, depending on the
+# availabilty of SO_REUSEPORT_WORKS. In which case the port initialization
+# is done in the worker processes.
+
+hitch_hosts() {
+	SO_REUSEPORT_WORKS=false
+	if test "$(python -c 'import socket; print(1 if hasattr(socket, "SO_REUSEPORT") else 0)')" = "1"
```
Comment (klenan) [inline src/tests/hitch_test.sh]: So this pulls in Python as test dependency, are you sure?
A shorter and simpler idiom would be

    if python -c 'import socket; exit(not hasattr(socket, "SO_REUSEPORT"))'

or

    if ! python -c 'import socket; exit(hasattr(socket, "SO_REUSEPORT"))'

depending on your taste.

File: src/tests/hitch_test.sh [hunk 3]
```diff
@@ -252,6 +298,11 @@ curl_hitch() {
 	if ! $HAS_SPECIFIC_ARG
 	then
 		HITCH_HOST=$(hitch_hosts | sed 1q)
+		if [ -z $HITCH_HOST ];
```
Comment (klenan) [inline src/tests/hitch_test.sh]: `$HITCH_HOST` should probably be quoted here.

File: src/tests/hitch_test.sh [hunk 4]
```diff
@@ -318,6 +369,10 @@ s_client() {
 	if ! $HAS_CONNECT_OPT
 	then
 		HITCH_HOST=$(hitch_hosts | sed 1q)
+		if [ -z $HITCH_HOST ];
```
Comment (klenan) [inline src/tests/hitch_test.sh]: Sames as above.

**Comment by dmatetelki** [comment]
@daghf mentioned that this whole thing might not work on reload, i need to investigate.

**Comment by daghf** [comment]
I'm working on a fix that supersedes this one. Closing

**Pull request closed without merging**